sdk/iot/provisioning/src/az_iot_provisioning_client.c

```diff
@@ -47,161 +54,420 @@ AZ_NODISCARD az_result az_iot_provisioning_client_init(
+    {
+      AZ_RETURN_IF_FAILED(az_json_parser_skip_children(jp, tm->token));
+    }
+    else
+    {
+      (void)az_iot_provisioning_client_payload_error_code_parse(tm, out_state);
+    }
+  }
+
+  return AZ_OK;
+}
+
+AZ_INLINE az_result az_iot_provisioning_client_parse_payload(
+    az_span received_payload,
+    az_iot_provisioning_client_register_response* out_response)
+{
+  // Parse the payload:
+  az_json_parser jp;
+  az_json_token_member tm;
+  bool found_operation_id = false;
+  bool found_registration_state = false;
+  bool found_error = false;
+
+  AZ_RETURN_IF_FAILED(az_json_parser_init(&jp, received_payload));
+  AZ_RETURN_IF_FAILED(az_json_parser_parse_token(&jp, &tm.token));
+  if (tm.token.kind != AZ_JSON_TOKEN_OBJECT_START)
+  {
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
+  }
+
+  out_response->registration_information = registration_state_default();
+
+  while (az_succeeded(az_json_parser_parse_token_member(&jp, &tm)))
+  {
+    if (az_span_is_content_equal(AZ_SPAN_FROM_STR("operationId"), tm.name))
+    {
+      found_operation_id = true;
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(&tm.token, &out_response->operation_id));
+    }
+    else if (az_span_is_content_equal(AZ_SPAN_FROM_STR("status"), tm.name))
+    {
+      found_registration_state = true;
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(&tm.token, &out_response->registration_state));
+    }
+    else if (az_span_is_content_equal(AZ_SPAN_FROM_STR("registrationState"), tm.name))
+    {
+      AZ_RETURN_IF_FAILED(az_iot_provisioning_client_payload_registration_state_parse(
+          &jp, &tm, &out_response->registration_information));
+    }
+    else if (az_span_is_content_equal(AZ_SPAN_FROM_STR("trackingId"), tm.name))
+    {
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(
+          &tm.token, &out_response->registration_information.error_tracking_id));
+    }
+    else if (az_span_is_content_equal(AZ_SPAN_FROM_STR("message"), tm.name))
+    {
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(
+          &tm.token, &out_response->registration_information.error_message));
+    }
+    else if (az_span_is_content_equal(AZ_SPAN_FROM_STR("timestampUtc"), tm.name))
+    {
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(
+          &tm.token, &out_response->registration_information.error_timestamp));
+    }
+    else if (tm.token.kind == AZ_JSON_TOKEN_OBJECT_START)
+    {
+      AZ_RETURN_IF_FAILED(az_json_parser_skip_children(&jp, tm.token));
+    }
+    else if (az_succeeded(az_iot_provisioning_client_payload_error_code_parse(
+                 &tm, &out_response->registration_information)))
+    {
+      found_error = true;
+    }
+
+    // else ignore token.
+  }
+
+  if (!(found_registration_state && found_operation_id))
+  {
+    out_response->operation_id = AZ_SPAN_NULL;
+    out_response->registration_state = AZ_SPAN_FROM_STR("failed");
+
+    if (!found_error)
+    {
+      return AZ_ERROR_ITEM_NOT_FOUND;
```

Comment:
item not found or parsing error?

Reply:
See az_result.h - in this case I think the error makes sense as we didn't find the item we were looking for.
I don't think `AZ_ERROR_PARSER_UNEXPECTED_CHAR` is an option here because it actually parsed correctly but the required items were not found.